README.md
[![npm version](https://badge.fury.io/js/cordova-plugin-add-swift-support.svg)](https://badge.fury.io/js/cordova-plugin-add-swift-support) [![Build Status](https://travis-ci.org/akofman/cordova-plugin-add-swift-support.svg?branch=master)](https://travis-ci.org/akofman/cordova-plugin-add-swift-support) [![npm](https://img.shields.io/npm/dm/cordova-plugin-add-swift-support.svg)]()

# cordova-plugin-add-swift-support

![swift-128x128](https://cloud.githubusercontent.com/assets/579922/15999501/79196b48-3146-11e6-836e-061a7ef53571.png)

This [Cordova plugin](https://www.npmjs.com/package/cordova-plugin-add-swift-support) adds the Swift support to your iOS project.

## Installation

You can add this plugin directly to your project:

`cordova plugin add cordova-plugin-add-swift-support --save`

Or add it as a dependency into your own plugin:

`<dependency id="cordova-plugin-add-swift-support" version="2.0.2"/>`

By default, the Swift 4 support is added but the legacy version (2.3) can still be configured as a preference, inside the project's `config.xml`, within the `<platform name="ios">` section:

`<preference name="UseLegacySwiftLanguageVersion" value="true" />`

Or it is possible to specify the version as following, inside the project's `config.xml`, within the `<platform name="ios">` section:

`<preference name="UseSwiftLanguageVersion" value="5" />`

If needed, add a prefixed Bridging-Header file in your plugin in order to import frameworks (MyPlugin-Bridging-Header.h for instance).
As an example you can have a look at this [plugin](https://github.com/akofman/cordova-plugin-permissionScope).

If the `cordova-plugin-add-swift-support` plugin is already installed to your project, then you can add your own Swift plugin as usual, its prefixed Bridging-Header will be automatically found and merged.

If you have some targets in your project that don't need support for a bridging header (e.g. watch app written completely in Swift with no Objective-C dependencies), you can configure excluded targets in a comma-separated list as follows:

`<preference name="SwiftOnlyTargets" value="WatchExtension,TodayExtension" />`

## Contributing

The src folder contains ECMAScript 2015 source files, the minimum Node.js version is `6` (Boron).

## License

MIT


CHANGELOG.md
# Change Log

## [2.0.2-dev-1.0](https://github.com/akofman/cordova-plugin-add-swift-support/tree/2.0.2-dev-1.0) (2021-07-27)
[Full Changelog](https://github.com/akofman/cordova-plugin-add-swift-support/compare/2.0.1...2.0.2-dev-1.0)

**Merged pull requests:**

- every plugin can change project.pbxproj file,
- async operations can be exist with random queue, (promise/callback and etc.)
- every part with xcode.project should be existed only in one task (micro)


## [2.0.2](https://github.com/akofman/cordova-plugin-add-swift-support/tree/2.0.2) (2019-04-05)
[Full Changelog](https://github.com/akofman/cordova-plugin-add-swift-support/compare/2.0.1...2.0.2)

**Merged pull requests:**

- Script needs to be quoted [\#59](https://github.com/akofman/cordova-plugin-add-swift-support/pull/59) ([shivan](https://github.com/shivan))

## [2.0.1](https://github.com/akofman/cordova-plugin-add-swift-support/tree/2.0.1) (2019-04-01)
[Full Changelog](https://github.com/akofman/cordova-plugin-add-swift-support/compare/2.0.0...2.0.1)

**Closed issues:**

- add-swift-support.js of v2.0 is missing on npm [\#58](https://github.com/akofman/cordova-plugin-add-swift-support/issues/58)
- Support cordova-cli 9.0.0 - requireCordovaModule deprecation [\#56](https://github.com/akofman/cordova-plugin-add-swift-support/issues/56)
- Swift 4 support [\#55](https://github.com/akofman/cordova-plugin-add-swift-support/issues/55)
- Swift 4 is failing with latest cordova [\#50](https://github.com/akofman/cordova-plugin-add-swift-support/issues/50)

## [2.0.0](https://github.com/akofman/cordova-plugin-add-swift-support/tree/2.0.0) (2019-04-01)
[Full Changelog](https://github.com/akofman/cordova-plugin-add-swift-support/compare/1.7.2...2.0.0)

**Merged pull requests:**

- refactor: update tooling, add support for Cordova v9 [\#57](https://github.com/akofman/cordova-plugin-add-swift-support/pull/57) ([timbru31](https://github.com/timbru31))
- Document where to specify Swift version preference [\#53](https://github.com/akofman/cordova-plugin-add-swift-support/pull/53) ([eladnava](https://github.com/eladnava))

## [1.7.2](https://github.com/akofman/cordova-plugin-add-swift-support/tree/1.7.2) (2018-03-30)
[Full Changelog](https://github.com/akofman/cordova-plugin-add-swift-support/compare/1.7.1...1.7.2)

**Closed issues:**

- Implement catching promise rejections [\#51](https://github.com/akofman/cordova-plugin-add-swift-support/issues/51)
- \*-Briding-Header.h not used [\#48](https://github.com/akofman/cordova-plugin-add-swift-support/issues/48)
- SWIFT\_VERSION build setting not being applied @1.7.1 [\#47](https://github.com/akofman/cordova-plugin-add-swift-support/issues/47)
- Unhandled promise rejection \(rejection id: 1\): Error: Cannot find module '../cordova/platform\_metadata' [\#44](https://github.com/akofman/cordova-plugin-add-swift-support/issues/44)
- Can not restore this plugin when using cordova-cli@8.0.0  [\#43](https://github.com/akofman/cordova-plugin-add-swift-support/issues/43)
- Hook won't trigger [\#40](https://github.com/akofman/cordova-plugin-add-swift-support/issues/40)
- Error adding cordova-plugin-add-swift-support [\#39](https://github.com/akofman/cordova-plugin-add-swift-support/issues/39)
- OSX support [\#19](https://github.com/akofman/cordova-plugin-add-swift-support/issues/19)

**Merged pull requests:**

- Fix for cordova projects located in paths with spaces [\#52](https://github.com/akofman/cordova-plugin-add-swift-support/pull/52) ([TheLiquidSky](https://github.com/TheLiquidSky))

## [1.7.1](https://github.com/akofman/cordova-plugin-add-swift-support/tree/1.7.1) (2017-12-22)
[Full Changelog](https://github.com/akofman/cordova-plugin-add-swift-support/compare/1.7.0...1.7.1)

**Closed issues:**

- SWIFT\_VERSION build setting not being applied [\#42](https://github.com/akofman/cordova-plugin-add-swift-support/issues/42)
- Node.JS 8.X Support [\#38](https://github.com/akofman/cordova-plugin-add-swift-support/issues/38)
- Missing tag and changelog for 1.7.0 [\#33](https://github.com/akofman/cordova-plugin-add-swift-support/issues/33)

**Merged pull requests:**

- fix\(\): platform\_metadata is now project\_metadata [\#45](https://github.com/akofman/cordova-plugin-add-swift-support/pull/45) ([mhartington](https://github.com/mhartington))
- add OSX paltform support [\#37](https://github.com/akofman/cordova-plugin-add-swift-support/pull/37) ([tryadelion](https://github.com/tryadelion))

## [1.7.0](https://github.com/akofman/cordova-plugin-add-swift-support/tree/1.7.0) (2017-07-07)
[Full Changelog](https://github.com/akofman/cordova-plugin-add-swift-support/compare/1.6.0...1.7.0)

**Closed issues:**

- Wrong text merged in Briding-Header.h [\#31](https://github.com/akofman/cordova-plugin-add-swift-support/issues/31)
- Phonegap Build Support [\#28](https://github.com/akofman/cordova-plugin-add-swift-support/issues/28)
- how to ? [\#26](https://github.com/akofman/cordova-plugin-add-swift-support/issues/26)
- swift3 required me to explicitly expose plugin method [\#23](https://github.com/akofman/cordova-plugin-add-swift-support/issues/23)
- Keep Xcode7 compatibility [\#20](https://github.com/akofman/cordova-plugin-add-swift-support/issues/20)

**Merged pull requests:**

- Set swift version in config.xml [\#30](https://github.com/akofman/cordova-plugin-add-swift-support/pull/30) ([NiklasMerz](https://github.com/NiklasMerz))

## [1.6.0](https://github.com/akofman/cordova-plugin-add-swift-support/tree/1.6.0) (2017-01-10)
[Full Changelog](https://github.com/akofman/cordova-plugin-add-swift-support/compare/1.4.0...1.6.0)

**Closed issues:**

- Differences between plugins [\#24](https://github.com/akofman/cordova-plugin-add-swift-support/issues/24)
- Error: undefined is not a function [\#22](https://github.com/akofman/cordova-plugin-add-swift-support/issues/22)
- Debug builds should be compiled without optimizations [\#18](https://github.com/akofman/cordova-plugin-add-swift-support/issues/18)

src/add-swift-support.js
/*
* This hook adds all the needed config to implement a Cordova plugin with Swift.
*
*  - It adds a Bridging header importing Cordova/CDV.h if it's not already
*    the case. Else it concats all the bridging headers in one single file.
*
*    /!\ Please be sure not naming your bridging header file 'Bridging-Header.h'
*    else it won't be supported.
*
*  - It puts the ios deployment target to 7.0 in case your project would have a
*    lesser one.
*
*  - It updates the ALWAYS_EMBED_SWIFT_STANDARD_LIBRARIES build setting to YES.
*
*  - It updates the SWIFT_VERSION to 4.0.
*/

const fs = require('fs');
const path = require('path');
const xcode = require('xcode');
const childProcess = require('child_process');
const semver = require('semver');
const glob = require('glob');

module.exports = context => {
  const projectRoot = context.opts.projectRoot;

  // This script has to be executed depending on the command line arguments, not
  // on the hook execution cycle.
  if ((context.hook === 'after_platform_add' && context.cmdLine.includes('platform add')) ||
    (context.hook === 'after_prepare' && context.cmdLine.includes('prepare')) ||
    (context.hook === 'after_plugin_add' && context.cmdLine.includes('plugin add'))) {
    getPlatformVersionsFromFileSystem(context, projectRoot).then(platformVersions => {
      const IOS_MIN_DEPLOYMENT_TARGET = '7.0';
      const platformPath = path.join(projectRoot, 'platforms', 'ios');
      const config = getConfigParser(context, path.join(projectRoot, 'config.xml'));

      let bridgingHeaderPath;
      let bridgingHeaderContent;
      let projectName;
      let projectPath;
      let pluginsPath;
      let iosPlatformVersion;
      let pbxprojPath;
      let xcodeProject;

      const COMMENT_KEY = /_comment$/;
      let buildConfigs;
      let buildConfig;
      let configName;

      platformVersions.forEach((platformVersion) => {
        if (platformVersion.platform === 'ios') {
          iosPlatformVersion = platformVersion.version;
        }
      });

      if (!iosPlatformVersion) {
        return;
      }

      projectName = config.name();
      projectPath = path.join(platformPath, projectName);
      pbxprojPath = path.join(platformPath, projectName + '.xcodeproj', 'project.pbxproj');
      xcodeProject = xcode.project(pbxprojPath);
      pluginsPath = path.join(projectPath, 'Plugins');

      xcodeProject.parseSync();

      bridgingHeaderPath = getBridgingHeaderPath(projectPath, iosPlatformVersion);

      try {
        fs.statSync(bridgingHeaderPath);
      } catch (err) {
        // If the bridging header doesn't exist, we create it with the minimum
        // Cordova/CDV.h import.
        bridgingHeaderContent = ['//',
          '//  Use this file to import your target\'s public headers that you would like to expose to Swift.',
          '//',
          '#import <Cordova/CDV.h>'];
        fs.writeFileSync(bridgingHeaderPath, bridgingHeaderContent.join('\n'), { encoding: 'utf-8', flag: 'w' });
        xcodeProject.addHeaderFile('Bridging-Header.h');
      }

      buildConfigs = xcodeProject.pbxXCBuildConfigurationSection();

      const bridgingHeaderProperty = '"$(PROJECT_DIR)/$(PROJECT_NAME)' + bridgingHeaderPath.split(projectPath)[1] + '"';

      for (configName in buildConfigs) {
        if (!COMMENT_KEY.test(configName)) {
          buildConfig = buildConfigs[configName];
          if (xcodeProject.getBuildProperty('SWIFT_OBJC_BRIDGING_HEADER', buildConfig.name) !== bridgingHeaderProperty) {
            xcodeProject.updateBuildProperty('SWIFT_OBJC_BRIDGING_HEADER', bridgingHeaderProperty, buildConfig.name);
            console.log('Update IOS build setting SWIFT_OBJC_BRIDGING_HEADER to:', bridgingHeaderProperty, 'for build configuration', buildConfig.name);
          }
        }
      }

      // Look for any bridging header defined in the plugin
      const files = glob.sync('**/*Bridging-Header*.h', { cwd: pluginsPath });
      const bridgingHeader = path.basename(bridgingHeaderPath);
      const headers = files.map((filePath) => path.basename(filePath));

      // if other bridging headers are found, they are imported in the
      // one already configured in the project.
      let content = fs.readFileSync(bridgingHeaderPath, 'utf-8');

      headers.forEach((header) => {
        if (header !== bridgingHeader && !~content.indexOf(header)) {
          if (content.charAt(content.length - 1) !== '\n') {
            content += '\n';
          }
          content += '#import "' + header + '"\n';
          console.log('Importing', header, 'into', bridgingHeaderPath);
        }
      });
      fs.writeFileSync(bridgingHeaderPath, content, 'utf-8');

      for (configName in buildConfigs) {
        if (!COMMENT_KEY.test(configName)) {
          buildConfig = buildConfigs[configName];
          if (parseFloat(xcodeProject.getBuildProperty('IPHONEOS_DEPLOYMENT_TARGET', buildConfig.name)) < parseFloat(IOS_MIN_DEPLOYMENT_TARGET)) {
            xcodeProject.updateBuildProperty('IPHONEOS_DEPLOYMENT_TARGET', IOS_MIN_DEPLOYMENT_TARGET, buildConfig.name);
            console.log('Update IOS project deployment target to:', IOS_MIN_DEPLOYMENT_TARGET, 'for build configuration', buildConfig.name);
          }

          if (xcodeProject.getBuildProperty('ALWAYS_EMBED_SWIFT_STANDARD_LIBRARIES', buildConfig.name) !== 'YES') {
            xcodeProject.updateBuildProperty('ALWAYS_EMBED_SWIFT_STANDARD_LIBRARIES', 'YES', buildConfig.name);
            console.log('Update IOS build setting ALWAYS_EMBED_SWIFT_STANDARD_LIBRARIES to: YES', 'for build configuration', buildConfig.name);
          }

          if (xcodeProject.getBuildProperty('LD_RUNPATH_SEARCH_PATHS', buildConfig.name) !== '"@executable_path/Frameworks"') {
            xcodeProject.updateBuildProperty('LD_RUNPATH_SEARCH_PATHS', '"@executable_path/Frameworks"', buildConfig.name);
            console.log('Update IOS build setting LD_RUNPATH_SEARCH_PATHS to: @executable_path/Frameworks', 'for build configuration', buildConfig.name);
          }

          if (typeof xcodeProject.getBuildProperty('SWIFT_VERSION', buildConfig.name) === 'undefined') {
            if (config.getPreference('UseLegacySwiftLanguageVersion', 'ios')) {
              xcodeProject.updateBuildProperty('SWIFT_VERSION', '2.3', buildConfig.name);
              console.log('Use legacy Swift language version', buildConfig.name);
            } else if (config.getPreference('UseSwiftLanguageVersion', 'ios')) {
              const swiftVersion = config.getPreference('UseSwiftLanguageVersion', 'ios');
              xcodeProject.updateBuildProperty('SWIFT_VERSION', swiftVersion, buildConfig.name);
              console.log('Use Swift language version', swiftVersion);
            } else {
              xcodeProject.updateBuildProperty('SWIFT_VERSION', '4.0', buildConfig.name);
              console.log('Update SWIFT version to 4.0', buildConfig.name);
            }
          }

          if (buildConfig.name === 'Debug') {
            if (xcodeProject.getBuildProperty('SWIFT_OPTIMIZATION_LEVEL', buildConfig.name) !== '"-Onone"') {
              xcodeProject.updateBuildProperty('SWIFT_OPTIMIZATION_LEVEL', '"-Onone"', buildConfig.name);
              console.log('Update IOS build setting SWIFT_OPTIMIZATION_LEVEL to: -Onone', 'for build configuration', buildConfig.name);
            }
          }
        }
      }

      fs.writeFileSync(pbxprojPath, xcodeProject.writeSync());
    })
      .catch(error => {
        console.error('add-swift-support.js error: ' + JSON.stringify(error));
        throw new Error(error);
      });
  }
};

const getConfigParser = (context, configPath) => {
  let ConfigParser;

  if (semver.lt(context.opts.cordova.version, '5.4.0')) {
    ConfigParser = context.requireCordovaModule('cordova-lib/src/ConfigParser/ConfigParser');
  } else {
    ConfigParser = context.requireCordovaModule('cordova-common/src/ConfigParser/ConfigParser');
  }

  return new ConfigParser(configPath);
};

const getBridgingHeaderPath = (projectPath, iosPlatformVersion) => {
  let bridgingHeaderPath;
  if (semver.lt(iosPlatformVersion, '4.0.0')) {
    bridgingHeaderPath = path.posix.join(projectPath, 'Plugins', 'Bridging-Header.h');
  } else {
    bridgingHeaderPath = path.posix.join(projectPath, 'Bridging-Header.h');
  }

  return bridgingHeaderPath;
};

const getPlatformVersionsFromFileSystem = (context, projectRoot) => {
  const cordovaUtil = context.requireCordovaModule('cordova-lib/src/cordova/util');
  const platformsOnFs = cordovaUtil.listPlatforms(projectRoot);
  const platformVersions = platformsOnFs.map(platform => {
    const script = path.join(projectRoot, 'platforms', platform, 'cordova', 'version');
    return new Promise((resolve, reject) => {
      childProcess.exec('"' + script + '"', {}, (error, stdout, _) => {
        if (error) {
          reject(error);
          return;
        }
        resolve(stdout.trim());
      });
    }).then(result => {
      const version = result.replace(/\r?\n|\r/g, '');
      return { platform, version };
    }, (error) => {
      console.log(error);
      process.exit(1);
    });
  });

  return Promise.all(platformVersions);
};


src/main.js
/*
* This hook adds all the needed config to implement a Cordova plugin with Swift.
*
*  - It adds a Bridging header importing Cordova/CDV.h if it's not already
*    the case. Else it concats all the bridging headers in one single file.
*
*    /!\ Please be sure not naming your bridging header file 'Bridging-Header.h'
*    else it won't be supported.
*
*  - It puts the ios deployment target to 7.0 in case your project would have a
*    lesser one.
*
*  - It updates the ALWAYS_EMBED_SWIFT_STANDARD_LIBRARIES build setting to YES.
*
*  - It updates the SWIFT_VERSION to 3.0.
*/

import fs from 'fs';
import path from 'path';
import xcode from 'xcode';
import childProcess from 'child_process';

export default (context) => {
  const projectRoot = context.opts.projectRoot;
  const glob = context.requireCordovaModule('glob');

  // This script has to be executed depending on the command line arguments, not
  // on the hook execution cycle.
  if ((context.hook === 'after_platform_add' && context.cmdLine.includes('platform add')) ||
  (context.hook === 'after_prepare' && context.cmdLine.includes('prepare')) ||
  (context.hook === 'after_plugin_add' && context.cmdLine.includes('plugin add'))) {
    getPlatformVersionsFromFileSystem(context, projectRoot).then((platformVersions) => {
      const IOS_MIN_DEPLOYMENT_TARGET = '7.0';
      const platformPath = path.join(projectRoot, 'platforms', 'ios');
      const config = getConfigParser(context, path.join(projectRoot, 'config.xml'));

      let bridgingHeaderPath;
      let bridgingHeaderContent;
      let projectName;
      let projectPath;
      let pluginsPath;
      let iosPlatformVersion;
      let pbxprojPath;
      let xcodeProject;

      const COMMENT_KEY = /_comment$/;
      let buildConfigs;
      let buildConfig;
      let configName;

      platformVersions.forEach((platformVersion) => {
        if (platformVersion.platform === 'ios') {
          iosPlatformVersion = platformVersion.version;
        }
      });

      if (!iosPlatformVersion) {
        return;
      }

      projectName = config.name();
      projectPath = path.join(platformPath, projectName);
      pbxprojPath = path.join(platformPath, projectName + '.xcodeproj', 'project.pbxproj');
      xcodeProject = xcode.project(pbxprojPath);
      pluginsPath = path.join(projectPath, 'Plugins');

      xcodeProject.parseSync();

      bridgingHeaderPath = getBridgingHeaderPath(context, projectPath, iosPlatformVersion);

      try {
        fs.statSync(bridgingHeaderPath);
      } catch (err) {
        // If the bridging header doesn't exist, we create it with the minimum
        // Cordova/CDV.h import.
        bridgingHeaderContent = [ '//',
          '//  Use this file to import your target\'s public headers that you would like to expose to Swift.',
          '//',
          '#import <Cordova/CDV.h>' ];
        fs.writeFileSync(bridgingHeaderPath, bridgingHeaderContent.join('\n'), { encoding: 'utf-8', flag: 'w' });
        xcodeProject.addHeaderFile('Bridging-Header.h');
      }

      buildConfigs = xcodeProject.pbxXCBuildConfigurationSection();

      const bridgingHeaderProperty = '"$(PROJECT_DIR)/$(PROJECT_NAME)' + bridgingHeaderPath.split(projectPath)[1] + '"';

      for (configName in buildConfigs) {
        if (!COMMENT_KEY.test(configName)) {
          buildConfig = buildConfigs[configName];
          if (xcodeProject.getBuildProperty('SWIFT_OBJC_BRIDGING_HEADER', buildConfig.name) !== bridgingHeaderProperty) {
            xcodeProject.updateBuildProperty('SWIFT_OBJC_BRIDGING_HEADER', bridgingHeaderProperty, buildConfig.name);
            console.log('Update IOS build setting SWIFT_OBJC_BRIDGING_HEADER to:', bridgingHeaderProperty, 'for build configuration', buildConfig.name);
          }
        }
      }

      // Look for any bridging header defined in the plugin
      glob('**/*Bridging-Header*.h', { cwd: pluginsPath }, (error, files) => {
        const bridgingHeader = path.basename(bridgingHeaderPath);
        const headers = files.map((filePath) => {
          return path.basename(filePath);
        });

        // if other bridging headers are found, they are imported in the
        // one already configured in the project.
        let content = fs.readFileSync(bridgingHeaderPath, 'utf-8');

        if (error) throw new Error(error);

        headers.forEach((header) => {
          if (header !== bridgingHeader && !~content.indexOf(header)) {
            if (content.charAt(content.length - 1) !== '\n') {
              content += '\n';
            }
            content += '#import "' + header + '"\n';
            console.log('Importing', header, 'into', bridgingHeaderPath);
          }
        });
        fs.writeFileSync(bridgingHeaderPath, content, 'utf-8');

        for (configName in buildConfigs) {
          if (!COMMENT_KEY.test(configName)) {
            buildConfig = buildConfigs[configName];
            if (parseFloat(xcodeProject.getBuildProperty('IPHONEOS_DEPLOYMENT_TARGET', buildConfig.name)) < parseFloat(IOS_MIN_DEPLOYMENT_TARGET)) {
              xcodeProject.updateBuildProperty('IPHONEOS_DEPLOYMENT_TARGET', IOS_MIN_DEPLOYMENT_TARGET, buildConfig.name);
              console.log('Update IOS project deployment target to:', IOS_MIN_DEPLOYMENT_TARGET, 'for build configuration', buildConfig.name);
            }

            if (xcodeProject.getBuildProperty('ALWAYS_EMBED_SWIFT_STANDARD_LIBRARIES', buildConfig.name) !== 'YES') {
              xcodeProject.updateBuildProperty('ALWAYS_EMBED_SWIFT_STANDARD_LIBRARIES', 'YES', buildConfig.name);
              console.log('Update IOS build setting ALWAYS_EMBED_SWIFT_STANDARD_LIBRARIES to: YES', 'for build configuration', buildConfig.name);
            }

            if (xcodeProject.getBuildProperty('LD_RUNPATH_SEARCH_PATHS', buildConfig.name) !== '"@executable_path/Frameworks"') {
              xcodeProject.updateBuildProperty('LD_RUNPATH_SEARCH_PATHS', '"@executable_path/Frameworks"', buildConfig.name);
              console.log('Update IOS build setting LD_RUNPATH_SEARCH_PATHS to: @executable_path/Frameworks', 'for build configuration', buildConfig.name);
            }

            if (typeof xcodeProject.getBuildProperty('SWIFT_VERSION', buildConfig.name) === 'undefined') {
              if (config.getPreference('UseLegacySwiftLanguageVersion', 'ios')) {
                xcodeProject.updateBuildProperty('SWIFT_VERSION', '2.3', buildConfig.name);
                console.log('Use legacy Swift language version', buildConfig.name);
              } else if (config.getPreference('UseSwiftLanguageVersion', 'ios')) {
                const swiftVersion = config.getPreference('UseSwiftLanguageVersion', 'ios');
                xcodeProject.updateBuildProperty('SWIFT_VERSION', swiftVersion, buildConfig.name);
                console.log('Use Swift language version', swiftVersion);
              } else {
                xcodeProject.updateBuildProperty('SWIFT_VERSION', '3.0', buildConfig.name);
                console.log('Update SWIFT version to 3.0', buildConfig.name);
              }
            }

            if (buildConfig.name === 'Debug') {
              if (xcodeProject.getBuildProperty('SWIFT_OPTIMIZATION_LEVEL', buildConfig.name) !== '"-Onone"') {
                xcodeProject.updateBuildProperty('SWIFT_OPTIMIZATION_LEVEL', '"-Onone"', buildConfig.name);
                console.log('Update IOS build setting SWIFT_OPTIMIZATION_LEVEL to: -Onone', 'for build configuration', buildConfig.name);
              }
            }
          }
        }

        fs.writeFileSync(pbxprojPath, xcodeProject.writeSync());
      });
    });
  }
};

const getConfigParser = (context, configPath) => {
  const semver = context.requireCordovaModule('semver');
  let ConfigParser;

  if (semver.lt(context.opts.cordova.version, '5.4.0')) {
    ConfigParser = context.requireCordovaModule('cordova-lib/src/ConfigParser/ConfigParser');
  } else {
    ConfigParser = context.requireCordovaModule('cordova-common/src/ConfigParser/ConfigParser');
  }

  return new ConfigParser(configPath);
};

const getBridgingHeaderPath = (context, projectPath, iosPlatformVersion) => {
  const semver = context.requireCordovaModule('semver');
  let bridgingHeaderPath;
  if (semver.lt(iosPlatformVersion, '4.0.0')) {
    bridgingHeaderPath = path.posix.join(projectPath, 'Plugins', 'Bridging-Header.h');
  } else {
    bridgingHeaderPath = path.posix.join(projectPath, 'Bridging-Header.h');
  }

  return bridgingHeaderPath;
};

const getPlatformVersionsFromFileSystem = (context, projectRoot) => {
  const cordovaUtil = context.requireCordovaModule('cordova-lib/src/cordova/util');
  const Q = context.requireCordovaModule('q');
  const platformsOnFs = cordovaUtil.listPlatforms(projectRoot);
  const platformVersions = platformsOnFs.map((platform) => {
    const script = path.join(projectRoot, 'platforms', platform, 'cordova', 'version');
    return Q.ninvoke(childProcess, 'exec', '"' + script + '"', {}).then((result) => {
      const version = result[0];
      const versionCleaned = version.replace(/\r?\n|\r/g, '');
      return {platform: platform, version: versionCleaned};
    }, (error) => {
      console.log(error);
      process.exit(1);
    });
  });

  return Q.all(platformVersions);
};
